src/enzyme_ad/jax/Passes/EnzymeHLOOpt.cpp

```diff
@@ -5999,28 +5999,61 @@ struct PowSimplify
 
   LogicalResult matchAndRewriteImpl(stablehlo::PowOp op,
                                     PatternRewriter &rewriter) const {
+    if (isa<FloatType>(op.getType().getElementType())) {
+      DenseFPElementsAttr rhs;
+      if (matchPattern(op.getRhs(), m_Constant(&rhs))) {
+        bool allHalf = true, allOne = true, allNegOne = true, allZero = true,
+             allNegHalf = true, allTwo = true;
+        for (auto v : rhs) {
+          allHalf &= v.isExactlyValue(0.5);
+          allOne &= v.isExactlyValue(1.0);
+          allNegOne &= v.isExactlyValue(-1.0);
+          allZero &= v.isExactlyValue(0.0);
+          allNegHalf &= v.isExactlyValue(-0.5);
+          allTwo &= v.isExactlyValue(2.0);
+        }
 
-    SmallVector<Attribute> constants;
-    constants.assign(op->getNumOperands(), Attribute());
-    for (unsigned i = 0, e = op->getNumOperands(); i != e; ++i)
-      matchPattern(op->getOperand(i), m_Constant(&constants[i]));
+        // pow(X, -1) -> 1 / X
+        if (allNegOne) {
+          rewriter.replaceOpWithNewOp<stablehlo::DivOp>(
+              op,
+              rewriter.create<stablehlo::ConstantOp>(
+                  op.getLoc(), op.getType(),
+                  cast<ElementsAttr>(makeAttr(op.getType(), 1))),
+              op.getLhs());
+          return success();
+        }
 
-    if (isa<FloatType>(op.getType().getElementType())) {
-      // pow(X, 0.5) -> sqrt(X)
-      {
-        DenseFPElementsAttr rhs;
-        if (matchPattern(op.getRhs(), m_Constant(&rhs))) {
-          bool allHalf = true;
-          for (auto v : rhs) {
-            if (!v.isExactlyValue(0.5)) {
-              allHalf = false;
-              break;
-            }
-          }
-          if (allHalf) {
-            rewriter.replaceOpWithNewOp<stablehlo::SqrtOp>(op, op.getLhs());
-            return success();
-          }
+        // pow(X, -0.5) -> rsqrt(X)
+        if (allNegHalf) {
+          rewriter.replaceOpWithNewOp<stablehlo::RsqrtOp>(op, op.getLhs());
+          return success();
+        }
+
+        // pow(X, 0) -> 1
+        if (allZero) {
+          rewriter.replaceOpWithNewOp<stablehlo::ConstantOp>(
+              op, op.getType(), cast<ElementsAttr>(makeAttr(op.getType(), 1)));
```

Comment:
We also may want an opt for a base of zero or one

Reply:
I need to check, I would’ve expected 0^0 to be nan, and nan^0 to be nan

Reply:
https://en.m.wikipedia.org/wiki/Zero_to_the_power_of_zero Yeah okay leave it as one then, we should do a check on the zero base then

Reply:
Ah I thought this was in response to my Nan comment, I meant in addition to the exponent special cases we should also add base special cases too

Reply:
separated some div patterns into no-nan as well

Reply:
It might be worth making an explicit isnanop which we lower but you can have one: https://github.com/EnzymeAD/Enzyme-JAX/blob/ba42c5b203ccefd9b1dfe82ffc276d135d5b59c4/src/enzyme_ad/jax/Passes/ArithRaising.cpp#L263

That said I don't think we'd want to remove all the nan check stuff (but add a branch if not enabled)